File: leetcode/Arrays/leetcode-31/31.go
package arrays

func nextPermutation(nums []int) {
	if nums == nil || len(nums) == 0 {
		return
	}

	firstSmall := -1

	for i := len(nums) - 2; i >= 0; i-- {
		if nums[i] < nums[i+1] {
			firstSmall = i
			break
		}
	}

	if firstSmall == -1 {
		reverse(nums, 0, len(nums)-1)
		return
	}

	for i := len(nums) - 1; i > firstSmall; i-- {
		if nums[i] > nums[firstSmall] {
			nums[i], nums[firstSmall] = nums[firstSmall], nums[i]
			break
		}
	}

	reverse(nums, firstSmall+1, len(nums)-1)
	return
}

func reverse(nums []int, x, y int) {
	for x < y {
		nums[x], nums[y] = nums[y], nums[x]
		x++
		y--
	}
}

File: leetcode/Arrays/leetcode-31/31.java
class Solution {
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length == 0) {
            return;
        }

        int firstSmall = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                firstSmall = i;
                break;
            }
        }

        if (firstSmall == -1) {
            reverse(nums, 0, nums.length - 1);
            return;
        }

        for (int i = nums.length - 1; i > firstSmall; i--) {
            if (nums[i] > nums[firstSmall]) {
                swap(nums, i, firstSmall);
                break;
            }
        }

        reverse(nums, firstSmall + 1, nums.length - 1);
        return;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start++, end--);
        }
    }
}